frame/liquidity-bootstrapping/src/lib.rs

```diff
@@ -0,0 +1,725 @@
+#![cfg_attr(
+	not(test),
+	warn(
+		clippy::disallowed_method,
+		clippy::disallowed_type,
+		clippy::indexing_slicing,
+		clippy::todo,
+		clippy::unwrap_used,
+		clippy::panic
+	)
+)]
+#![warn(clippy::unseparated_literal_suffix)]
+#![cfg_attr(not(feature = "std"), no_std)]
+#![warn(
+	bad_style,
+	bare_trait_objects,
+	const_err,
+	improper_ctypes,
+	non_shorthand_field_patterns,
+	no_mangle_generic_items,
+	overflowing_literals,
+	path_statements,
+	patterns_in_fns_without_body,
+	private_in_public,
+	unconditional_recursion,
+	unused_allocation,
+	unused_comparisons,
+	unused_parens,
+	while_true,
+	trivial_casts,
+	trivial_numeric_casts,
+	unused_extern_crates
+)]
+
+pub use pallet::*;
+
+pub mod weights;
+
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarking;
+
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+mod tests;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use crate::weights::WeightInfo;
+	use codec::{Codec, FullCodec};
+	use composable_support::validation::{Validate, Validated};
+	use composable_traits::{
+		currency::LocalAssets, defi::CurrencyPair, dex::CurveAmm, math::SafeArithmetic,
+	};
+	use core::fmt::Debug;
+	use frame_support::{
+		pallet_prelude::*,
+		traits::fungibles::{Inspect, Transfer},
+		transactional, PalletId, RuntimeDebug,
+	};
+	use frame_system::{
+		ensure_signed,
+		pallet_prelude::{BlockNumberFor, OriginFor},
+	};
+	use rust_decimal::prelude::*;
+	use sp_arithmetic::traits::Saturating;
+	use sp_runtime::{
+		traits::{
+			AccountIdConversion, BlockNumberProvider, CheckedMul, CheckedSub, Convert, One, Zero,
+		},
+		ArithmeticError, Permill,
+	};
+
+	#[derive(Copy, Clone, PartialEq, Eq)]
+	pub enum SaleState {
+		NotStarted,
+		Ongoing,
+		Ended,
+	}
+
+	#[derive(RuntimeDebug, Encode, Decode, MaxEncodedLen, Copy, Clone, PartialEq, Eq, TypeInfo)]
+	pub struct Sale<BlockNumber> {
+		/// Block at which the sale start.
+		pub start: BlockNumber,
+		/// Block at which the sale stop.
+		pub end: BlockNumber,
+		/// Initial weight of the base asset of the current pair.
+		pub initial_weight: Permill,
+		/// Final weight of the base asset of the current pair.
+		pub final_weight: Permill,
+	}
+
+	impl<BlockNumber: TryInto<u64> + Ord + Copy + Saturating + SafeArithmetic> Sale<BlockNumber> {
+		pub(crate) fn current_weights(
+			&self,
+			current_block: BlockNumber,
+		) -> Result<(Permill, Permill), DispatchError> {
+			/* NOTE(hussein-aitlahcen): currently only linear
+
+			   Assuming final_weight < initial_weight
```

Comment:
please leave comment that base weight diminished with time in docs (so what we sell becomes cheaper in quote)

Reply:
done!